torch/distributed/_tensor/README.md

```diff
@@ -1,3 +1,167 @@
+colwise_tensor = rowwise_tensor.redistribute(device_mesh, colwise_placement)
+replica_tensor = colwise_tensor.redistribute(device_mesh, replica_placement)
+
+```
+
+#### High level User Facing APIs
+
+Users can use DistributedTensor tensor constructors directly to create a distributed tensor (i.e. `distributed.ones/empty`), but for existing modules like nn.Linear that are already having torch.Tensor as parameters, how to make them distributed parameters? We offer a way to directly distribute a torch.Tensor and a module level APIs to directly distribute the module parameters. Below is the high level API we introduce:
+
+```python
+def distribute_tensor(tensor: torch.Tensor, device_mesh: DeviceMesh=None, placements: List[Placement]=None):
+    '''
+    distribute the tensor according to device_mesh and placements, `tensor` could be a "meta" tensor.
+    '''
+
+def distribute_module(
+    module: nn.Module,
+    device_mesh: DeviceMesh=None,
+    partition_fn: Callable[[str, nn.Module, DeviceMesh], ...]=None,
+    input_fn: Callable[...., None]=None,
+    output_fn: Callable[...., None]=None,
+):
+    '''
+    This function converts all module parameters to distributed tensor parameters according to the `partition_fn` specified.
+    It could also control the input/output of the module by specifying the `input_fn` and `output_fn`.
+    '''
+```
+
+#### High level API examples:
+
+```python
+def MyModule(nn.Module):
+    def __init__(self):
+        super.__init__()
+        self.fc1 = nn.Linear(8, 8)
+        self.fc2 = nn.Linear(8, 8)
+        self.relu = nn.ReLU()
+
+    def forward(self, input):
+        return self.relu(self.fc1(input) + self.fc2(input))
+
+mesh = DeviceMesh(device_type="cuda", [[0, 1], [2, 3]])
+
+def shard_params(mod_name, mod, mesh):
+    rowwise_placement = [Shard(0)]
+    def to_dist_tensor(t): return distribute_tensor(t, mesh, rowwise_placement)
+    mod._apply(to_dist_tensor)
+
+sharded_module = distribute_module(model, device_mesh, partition_fn=shard_params)
+
+def shard_fc(mod_name, mod, mesh):
+    rowwise_placement = [Shard(0)]
+    if mod_name == "fc1":
+        mod.weight = torch.nn.Parameter(distribute_tensor(mod.weight, mesh, rowwise_placement))
+
+sharded_module = distribute_module(model, device_mesh, partition_fn=shard_fc)
+```
+
+## Compiler and DistributedTensor
+
+DistributedTensor provides efficient solutions for cases like Tensor Parallelism. But when using the DTensor's replication in a data parallel fashion, it might become observably slow compared to our existing solutions like DDP/FSDP. This is mainly because existing solutions like DDP/FSDP could have the global view of entire model architecture, thus could optimize for data parallel specifically, i.e. collective fusion and computation overlap, etc. DistributedTensor itself is only a Tensor-like object and only knows its local computation operation, it does not know the subsequent operations that happened afterwards.
```

Comment:
nit: slow -> slower

Reply:
nit: "mainly because existing solutions like DDP/FSDP could have the global view of entire model architecture" -> "mainly because DDP/FSDP have a global view of the entire model architecture, "